crates/sqllib/src/dec.rs

```diff
@@ -0,0 +1,923 @@
+//! Operations on Decimal values
+
+use crate::{
+    cast_to_SqlDecimal_s, some_existing_operator, some_function2, some_operator,
+    some_polymorphic_function1, some_polymorphic_function2, unwrap_cast, HasOne, HasZero,
+    SqlString,
+};
+use dbsp::algebra::{MulByRef, OptionWeightType, F64};
+use dec::{Context, Decimal, ParseDecimalError, Rounding, TryFromDecimalError};
+use feldera_types::deserialize_without_context;
+use feldera_types::serde_with_context::{
+    serde_config::DecimalFormat, SerializeWithContext, SqlSerdeConfig,
+};
+use num::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Zero};
+use rkyv::{string::StringResolver, DeserializeUnsized, Fallible, SerializeUnsized};
+use serde::de::Error as DeserError;
+use serde::ser::Error as SerError;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_json::Number;
+use size_of::SizeOf;
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+    ops::{Add, AddAssign, Div, Mul, Neg, Sub},
+    sync::LazyLock,
+};
+
+const MAX_DIGITS: usize = 13;
+const DECIMAL_MAX_PRECISION: usize = 3 * MAX_DIGITS;
+
+pub type LargeDecimal = Decimal<MAX_DIGITS>;
+
+#[derive(Copy, Debug, Default, Clone)]
+pub struct SqlDecimal {
+    value: LargeDecimal,
+}
+
+pub(crate) type DecimalContext = Context<LargeDecimal>;
+
+#[doc(hidden)]
+pub(crate) fn is_error(context: &DecimalContext) -> bool {
+    let status = context.status();
+    status.insufficient_storage()
+        || status.overflow()
+        || status.invalid_operation()
+        || status.division_by_zero()
+        || status.division_undefined()
+        || status.division_impossible()
+        || status.underflow()
```

Comment:
I don't know whether underflow should be treated as an error rather than allowing it to become a subnormal or flush to zero.

Reply:
removed underflow from this list.